File: dec-crc32-cracker.c

#define WASM_EXPORT __attribute__((visibility("default")))
#define u32 unsigned int
#define i32 int
#define u8 unsigned char
#define i8 char
u32 table[256];
WASM_EXPORT void *init(u32 POLY)
{
  for (u32 i = 0u; i < 256u; i++)
  {
    u32 j = i;
    for (u8 _ = 0u; _ < 8u; _++)
    {
      j = j & 1u ? j >> 1u ^ POLY : j >> 1u;
    }
    table[i] = j;
  }
  return table;
}
u8 crc_index;
u32 crc32(u8 *str, u8 len)
{
  u32 crc = ~0u;
  for (u8 i = 0u; i < len; i++)
  {
    crc_index = (crc ^ str[i]) & 0xFFu;
    crc = crc >> 8u ^ table[crc_index];
  }
  return crc;
}
u8 lastIndex(u8 crc)
{
  for (u32 i = 0u; i < 256u; i++)
  {
    if (crc == table[i] >> 24u)
    {
      return i;
    }
  }
  return 0u;
}
union
{
  u32 value;
  u8 bytes[4];
} indexes;
WASM_EXPORT u32 getIndexes(u32 crc)
{
  for (u8 i = 3u, j = 0u; j < 4u; i--, j++)
  {
    u8 crc_index = indexes.bytes[j] = lastIndex(crc >> (i << 3u));
    u32 value = table[crc_index];
    crc ^= value >> (j << 3u);
  }
  return indexes.value;
}
u8 itoa(u32 unum, u8 *str)
{
  u8 i = 0u;
  do
  {
    str[i++] = unum % 10u + 48u;
    unum /= 10u;
  } while (unum);
  for (u8 j = 0u; j <= (i - 1u) / 2u; j++)
  {
    u8 temp = str[j];
    str[j] = str[i - 1u - j];
    str[i - 1u - j] = temp;
  }
  return i;
}
u8 str[10];
WASM_EXPORT u32 crc32_dec(u32 i)
{
  u8 len = itoa(i, str);
  u32 crc = crc32(str, len);
  return crc;
}
u32 check(u32 i)
{
  u8 len = itoa(i, str);
  u32 crc = crc32(str, len);
  if (crc_index != indexes.bytes[3])
  {
    return -1;
  }
  u32 low = 0u;
  for (i8 i = 2; i > -1; i--)
  {
    u8 num = (crc & 0xFFu) ^ indexes.bytes[i];
    if (num < 48u || num > 57u)
    {
      return -1;
    }
    low = low << 4u | (num - 48u);
    crc = crc >> 8u ^ table[indexes.bytes[i]];
  }
  return low;
}
WASM_EXPORT u32 checkIndex(u32 i, u32 indexes_value)
{
  indexes.value = indexes_value;
  return check(i);
}
WASM_EXPORT i32 findIndex(u32 indexes_value, i32 begin, i32 end)
{
  indexes.value = indexes_value;
  for (i32 i = begin; i < end; i++)
  {
    i32 low = check(i);
    if (low >= 0)
    {
      return i;
    }
  }
  return -1;
}

File: dec-crc32-cracker.zig
const zeroes = @import("std").mem.zeroes;

var table = zeroes([256]u32);
var lastIndex = zeroes([256]u8);
pub export fn init(POLY: u32) usize {
    for (0..256) |i| {
        var j: u32 = @intCast(i);
        for (0..8) |_| {
            j = if ((j & 1) != 0)
                (j >> 1) ^ POLY
            else
                j >> 1;
        }
        table[i] = j;
        lastIndex[j >> 24] = @intCast(i);
    }
    return @intFromPtr(&table);
}
pub var crc_index: u8 = 0;
pub fn crc32(buf: []const u8) u32 {
    var crc = ~@as(u32, 0);
    for (buf) |value| {
        crc_index = @truncate(crc ^ value);
        crc = (crc >> 8) ^ table[crc_index];
    }
    return crc;
}
const Indexes = extern union {
    value: u32,
    bytes: [4]u8,
};
pub var indexes = zeroes(Indexes);
pub export fn getIndexes(_crc: u32) u32 {
    var crc = _crc;
    var j: u5 = 0;
    while (j < 4) : (j += 1) {
        const i: u5 = 3 - j;
        const _crc_index = lastIndex[crc >> (i << 3)];
        indexes.bytes[j] = _crc_index;
        crc ^= table[_crc_index] >> (j << 3);
    }
    return indexes.value;
}

pub fn itoa(num: u32, buf: *[10]u8) []const u8 {
    var i: u8 = buf.len;
    var tmp = num;
    while (true) {
        i -= 1;
        buf[i] = @truncate((tmp % 10) + '0');
        tmp /= 10;
        if (tmp == 0) break;
    }
    return buf[i..];
}
pub var str = zeroes([10]u8);
pub export fn crc32_dec(i: u32) u32 {
    return crc32(itoa(i, &str));
}
pub fn check(_i: u32) i32 {
    var crc: u32 = crc32_dec(_i);
    if (crc_index != indexes.bytes[3]) {
        return -1;
    }
    var low: u12 = 0;
    var i: u8 = 2;
    while (true) : (i -= if (i == 0) break else 1) {
        const num: u8 = @as(u8, @truncate(crc)) ^ indexes.bytes[i];
        if (num < '0' or num > '9') {
            return -1;
        }
        low = (low << 4) | (num - '0');
        crc = (crc >> 8) ^ table[indexes.bytes[i]];
    }
    return low;
}
pub export fn checkIndex(i: i32, indexes_value: u32) i32 {
    indexes.value = indexes_value;
    return check(@intCast(i));
}
pub export fn findIndex(indexes_value: u32, begin: i32, end: i32) i32 {
    indexes.value = indexes_value;
    var i = begin;
    while (i < end) : (i += 1) {
        const low = check(@intCast(i));
        if (low >= 0) {
            return i;
        }
    }
    return -1;
}
